frigate: coral tpu support, audio model, nvidia ffmpeg hwaccel, other fixes

mweinelt opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->
Supersedes: #332689
Closes: #320512, #344114

cc @nagisa

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

mweinelt committed:
libedgetpu: use dedicated coral group

mweinelt committed:
nixos/coral: init

Provides a small wrapper to enable support for Coral USB and PCIe
devices.

mweinelt committed:
nixos/frigate: add support for Coral devices

mweinelt committed:
nixos/frigate: stop enabling recommendedProxySettings globally

Closes: #320512

mweinelt committed:
frigate: provide the tflite audio model

mweinelt committed:
nixos/frigate: provide ffmpeg-full for nvidia hw accel

Closes: #344114

mweinelt committed:
linuxPackages_latest.gasket: fix build with 6.12

nagisa reviewed (approved):
The nginx changes seem good to me. The gasket/edgetpu changes LGTM.

Can't comment on tensorflow/model/label or nvidia changes.

JManch commented:
I haven't tested this PR but on my system I need the following for the PCIe Coral to work as described [here](https://coral.ai/docs/m2/get-started/#requirements)
```nix
users.groups.apex = {};

services.udev.extraRules = ''
  SUBSYSTEM=="apex", MODE="0660", GROUP="apex"
'';

systemd.services.frigate.serviceConfig.SupplementaryGroups = "apex";
```

mweinelt commented:
Okay, renamed the edgetpu module to `hardware.coral` and also named the group for access to coral devices that.

@JManch Thanks for the pointer, I've integrates those changes.

netpleb commented:
Sorry, I had originally commented on the wrong issue (#188719). @mweinelt I am happy to give this a try. It is pretty much exactly the fix I have been looking for! I have a flakes based system with a nixpkgs input pointed at nixpkgs-unstable. How would you suggest is the best way for me to try out your PR and report back? Do I need to somehow disable the existing nixpkgs frigate module?

GoogleBot42 commented on nixos/modules/hardware/coral.nix:
```diff
@@ -0,0 +1,44 @@
+    (mkIf cfg.pcie.enable {
+      boot = {
+        kernelPackages = with config.boot.kernelPackages; [ gasket ];
+        kernelModules = [
```
I think preloading the kernel modules isn't necessary.  Or at least I don't need it for my coral pcie tpu.

GoogleBot42 commented:
This looks pretty good!  I'll see if I can give it a try later today.  It looks like it should be able to replace what I have already. https://github.com/GoogleBot42/nix-config/blob/eb6a50664cb3efbecde19cb9316a682cf64e983d/machines/storage/s0/frigate.nix#L143

mweinelt committed:
nixos/frigate: inherit required functions from lib

JManch commented on nixos/modules/hardware/coral.nix:
```diff
@@ -0,0 +1,38 @@
+      services.udev.packages = with pkgs; [ libedgeptu ];
+    })
+    (mkIf cfg.pcie.enable {
+      boot.kernelPackages = with config.boot.kernelPackages; [ gasket ];
```
```suggestion
      boot.extraModulePackages = with config.boot.kernelPackages; [ gasket ];
```

mweinelt replied:
Wow, silly me.

JManch commented:
Just tested my PCIe Coral with this PR and can confirm it works as expected!

JManch commented:
https://github.com/NixOS/nixpkgs/blob/44fe6539a5130376588ceb9425d3b69201d40193/nixos/modules/services/video/frigate.nix#L559
Just as an extra small fix could this be replaced with something like:
```nix
ExecStartPre = pkgs.writeShellScript "frigate-clear-cache" ''
  rm -f /var/cache/frigate/*.mp4
'';
```
Currently I get `frigate.service: Unable to locate executable 'rm': No such file or directory` every time the service starts and I don't think the wildcard expansion works in this context.

mweinelt committed:
nixos/frigate: use shellscript to clear frigate cache

Shell expansions apparently do not work correctly in systemd command
lines.

Co-Authored-By: Joshua Manchester <JManch@protonmail.com>

mweinelt commented on nixos/modules/services/video/frigate.nix:
```diff
@@ -94,6 +103,15 @@ let
+
+  # Provide ffmpeg-full for NVIDIA hardware acceleration
+  ffmpegArgs = cfg.settings.ffmpeg.hwaccel_args or "";
+  ffmpeg' = if match "/nvidia/" ffmpegArgs != null then pkgs.ffmpeg-full else pkgs.ffmpeg-headless;
```
cudaLLVM.

Atemu replied:
https://github.com/NixOS/nixpkgs/pull/354952 now fixes this.

mweinelt replied:
This gets backported, so unless the uplift also gets backported I can only make use of this later.

Atemu replied:
Theoretically, we could backport. I don't really care what happens on 24.11 w.r.t. ffmpeg-full though, so this would also be fine. I'd just revert this when the time comes to remove ffmpeg-full tree-wide.

mweinelt committed:
nixos/frigate: allow GPU use for video acceleration

The `render` group is required to acces the graphic cards decoding,
encoding and transcoding capabilities.

It is required for using fancy `ffmpeg.hwaccel-args` values, like
`preset-vaapi`.

mweinelt committed:
nixos/frigate: allow configuring a libva driver

In my testing this helped steer VA-API towards the correct card and also
made it show GPU statistics in performance monitoring.

JManch commented on nixos/modules/services/video/frigate.nix:
```diff
@@ -510,26 +551,35 @@ in
       serviceConfig = {
-        ExecStartPre = "-rm /var/cache/frigate/*.mp4";
+        ExecStartPre = pkgs.writeShellScript "frigate-clear-cache" ''
+          rm --force /var/cache/frigate/*.mp4
```
Sorry for the hassle but after looking at the intention of the original PR #337520, it probably makes more sense to do `rm --recursive --force /var/cache/frigate/*` since frigate also makes a `preview_frames` dir. Thanks!

frederictobiasc commented on nixos/modules/hardware/coral.nix:
```diff
@@ -0,0 +1,38 @@
+      users.groups.coral = { };
+    })
+    (mkIf cfg.usb.enable {
+      services.udev.packages = with pkgs; [ libedgeptu ];
```
```suggestion
      services.udev.packages = with pkgs; [ libedgetpu ];
```

Atemu commented:
Ideally we'd have https://github.com/NixOS/nixpkgs/issues/350162 where we could simply assert that a VAAPI driver is configured.

For now it'd be best if the option description had a `:::{.note}` that informs the user that they must ensure VAAPI works on their machine.

mweinelt commented:
> For now it'd be best if the option description had a :::{.note} that informs the user that they must ensure VAAPI works on their machine.

Added in the last commit, however the rendering of admonitions on s.n.o. is non-existent.

Atemu commented on nixos/modules/services/video/frigate.nix:
```diff
@@ -114,6 +132,20 @@ in
+      default = null;
+      example = "radeonsi";
+      description = ''
+        VA-API driver to use for video acceleration. Use together with `settings.ffmpeg.hwaccel_args`.
+
+        - https://docs.frigate.video/configuration/hardware_acceleration
+        - https://docs.frigate.video/configuration/ffmpeg_presets#hwaccel-presets
+
+        Requires a VA-API driver to be installed through `hardware.graphics.extraPackages`.
```
```suggestion
        Force usage of a particular VA-API driver for video acceleration. Use together with `settings.ffmpeg.hwaccel_args`.

        Setting this is not required for VA-API to work but it can help steer it towards the correct card if you have multiple.

        :::{.note}
        You must configure a VA-API driver package for VA-API to work, regardless of this setting.
        AMDGPUs require {option}`hardware.graphics.enable` and most Intel GPUs need `pkgs.intel-media-driver` in {option}`hardware.graphics.extraPackages`.
        :::

        See also:

        - https://docs.frigate.video/configuration/hardware_acceleration
        - https://docs.frigate.video/configuration/ffmpeg_presets#hwaccel-presets
```

mweinelt replied:
Again, are option descriptions rendered with admonitions *anywhere at all*?

Atemu replied:
It renders fine in the options manual:

    nix-build nixos/release.nix -A manual.x86_64-linux && xdg-open ./result/share/doc/nixos/options.html

Atemu replied:
I also just confirmed that it renders fine in `man configuration.nix` too.

s.n.o currently just strips the admonition text but keeps the content as if it was regular content. See i.e. https://search.nixos.org/options?channel=unstable&show=security.dhparams.params&from=0&size=50&sort=relevance&type=packages&query=security.dhparams.params where the last 3 paragraphs are note, warning and note.

Once it gets proper support, it should simply "just work".

Atemu commented:
> the rendering of admonitions on s.n.o. is non-existent

If it renders fine in the options manual, that's good enough IMHO. It will get the point across to s.n.o users, even if it's a little janky.

mweinelt committed:
frigate: patch path to birdseye graphic

mweinelt commented:
Last call everyone! If you have further feedback, send it now.

The plan is to merge this in later tonight (in ~6 hours).

netpleb commented:
@frederictobiasc I have been trying to test this with my coral USB and frigate but to no avail. Would it be possible for you to share the relevant parts of the configuration you used to test this PR?

netpleb commented:
Actually, a reboot seemed to fix the issue and now it properly detects the usb tpu it seems. I am still new to frigate so not totally able to test the detection functionality itself yet, but this was a huge step! Thanks so much @mweinelt !!!

mweinelt commented:
Hey everyone, feel free to join [#homeautomation:nixos.org](https://matrix.to/#/#homeautomation:nixos.org) on Matrix to stay connected. We talk a lot about Home Assistant, Frigate as well as Zigbee, Matter and Z-Wave in there.
